File: src/main/scala/com/xebia/functional/chains/BaseChain.scala
package com.xebia.functional.chains

import cats.MonadThrow
import cats.syntax.all.*

import com.xebia.functional.chains.models.Config

trait BaseChain[F[_]: MonadThrow]:

  val config: Config
  val inputVariables: Set[String] = config.inputKeys
  val outputVariables: Set[String] = config.outputKeys

  def call(inputs: Map[String, String]): F[Map[String, String]]

  private def prepareInputs(inputs: Map[String, String] | String): F[Map[String, String]] =
    inputs match
      case e: String => config.genInputsFromString(e)
      case e: Map[String, String] => config.genInputs(e)

  private def prepareOutputs(inputs: Map[String, String])(outputs: Map[String, String]): F[Map[String, String]] =
    MonadThrow[F].pure(
      config.onlyOutputs match
        case true => outputs
        case false => inputs ++ outputs
    )
  
  def run(inputs: Map[String, String] | String): F[Map[String, String]] = 
    prepareInputs(inputs).flatMap( is => call(is) >>= prepareOutputs(is))

File: src/commonMain/kotlin/com/xebia/functional/chains/BaseChain.kt
package com.xebia.functional.chains

import arrow.core.Either
import arrow.core.raise.either

interface BaseChain {

    val config: Config

    suspend fun call(inputs: Map<String, String>): Map<String, String>

    suspend fun run(input: String): Either<InvalidChainInputs, Map<String, String>> =
        either {
            val preparedInputs = config.genInputsFromString(input).bind()
            val result = call(preparedInputs)
            prepareOutputs(preparedInputs, result)
        }

    suspend fun run(inputs: Map<String, String>): Either<InvalidChainInputs, Map<String, String>> =
        either {
            val preparedInputs = config.genInputs(inputs).bind()
            val result = call(preparedInputs)
            prepareOutputs(preparedInputs, result)
        }

    private suspend fun prepareOutputs(
        inputs: Map<String, String>, outputs: Map<String, String>
    ): Map<String, String> =
        if (config.onlyOutputs) outputs else inputs + outputs
}
